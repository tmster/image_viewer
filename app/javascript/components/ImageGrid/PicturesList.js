import React from 'react';
import PropTypes from 'prop-types';
import InfiniteScroll from 'react-infinite-scroller';
import Row from './Row';

class PicturesList extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      fetchMore: props.fetchMore,
      hasMore: true,
      pictures: props.pictures,
    };
  }

  loadMore() {
    this.state.fetchMore({
      variables: {
        offset: this.state.pictures.length,
      },
      updateQuery: (prev, { fetchMoreResult }) => {
        if (!fetchMoreResult) return this.state.pictures;

        const pictures = [...this.state.pictures, ...fetchMoreResult.pictures];

        if (this.state.pictures.length === pictures.length) {
          return this.setState({ hasMore: false });
        }

        return this.setState({
          pictures,
        });
      },
    });
  }

  showItems() {
    const groups = this.groupedPictures();
    const items = groups.map((group, index) => (
      <Row key={index} group={group} />
    ));
    return items;
  }

  groupedPictures(n = 4) {
    const { pictures } = this.state;
    const group = [];
    for (let i = 0, end = pictures.length / n; i < end; i += 1)
      group.push(pictures.slice(i * n, (i + 1) * n));
    return group;
  }

  render() {
    return (
      <div className="container image-grid">
        <InfiniteScroll
          loadMore={this.loadMore.bind(this)}
          hasMore={this.state.hasMore}
          useWindow={true}
        >
          {this.showItems()}
        </InfiniteScroll>
      </div>
    );
  }
}

PicturesList.propTypes = {
  pictures: PropTypes.array.isRequired,
  fetchMore: PropTypes.func.isRequired,
};

export default PicturesList;
