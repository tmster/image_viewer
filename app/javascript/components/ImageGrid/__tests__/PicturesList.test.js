import React from 'react';
import { shallow } from 'enzyme';

import PicturesList from '../PicturesList';

const pictures = [
  { id: '1', imageUrl: 'http://localhost/1', title: 'Awesome' },
  { id: '2', imageUrl: 'http://localhost/2', title: 'Poor' },
];

const fetchMoreMock = () => {};

it('renders all props', () => {
  const node = shallow(
    <PicturesList pictures={pictures} fetchMore={fetchMoreMock} />
  );
  expect(node).toMatchSnapshot();
});

it('groups pictures', () => {
  const node = shallow(
    <PicturesList pictures={pictures} fetchMore={fetchMoreMock} />
  );

  expect(node.instance().groupedPictures(1)).toEqual([
    [pictures[0]],
    [pictures[1]],
  ]);

  expect(node).toMatchSnapshot();
});
