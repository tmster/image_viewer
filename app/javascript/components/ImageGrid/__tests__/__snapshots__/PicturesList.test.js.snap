// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groups pictures 1`] = `
<div
  className="container image-grid"
>
  <InfiniteScroll
    element="div"
    getScrollParent={null}
    hasMore={true}
    initialLoad={true}
    isReverse={false}
    loadMore={[Function]}
    loader={null}
    pageStart={0}
    ref={null}
    threshold={250}
    useCapture={false}
    useWindow={true}
  >
    <Row
      group={
        Array [
          Object {
            "id": "1",
            "imageUrl": "http://localhost/1",
            "title": "Awesome",
          },
          Object {
            "id": "2",
            "imageUrl": "http://localhost/2",
            "title": "Poor",
          },
        ]
      }
      key="0"
    />
  </InfiniteScroll>
</div>
`;

exports[`renders all props 1`] = `
<div
  className="container image-grid"
>
  <InfiniteScroll
    element="div"
    getScrollParent={null}
    hasMore={true}
    initialLoad={true}
    isReverse={false}
    loadMore={[Function]}
    loader={null}
    pageStart={0}
    ref={null}
    threshold={250}
    useCapture={false}
    useWindow={true}
  >
    <Row
      group={
        Array [
          Object {
            "id": "1",
            "imageUrl": "http://localhost/1",
            "title": "Awesome",
          },
          Object {
            "id": "2",
            "imageUrl": "http://localhost/2",
            "title": "Poor",
          },
        ]
      }
      key="0"
    />
  </InfiniteScroll>
</div>
`;
